src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1177 @@
+static bool IsUniformHashDistribution(ShardInterval **shardIntervalArray,
+									  int shardCount);
+static FmgrInfo * ShardIntervalCompareFunction(Var *partitionColumn, char
+											   partitionMethod);
+static ShardInterval * FindShardInterval(Datum partitionColumnValue,
+										 ShardInterval **shardIntervalCache,
+										 int shardCount, char partitionMethod,
+										 FmgrInfo *compareFunction,
+										 FmgrInfo *hashFunction, bool useBinarySearch);
+static ShardInterval * SearchCachedShardInterval(Datum partitionColumnValue,
+												 ShardInterval **shardIntervalCache,
+												 int shardCount,
+												 FmgrInfo *compareFunction);
+static void OpenCopyTransactions(CopyStmt *copyStatement,
+								 ShardConnections *shardConnections,
+								 int64 shardId);
+static StringInfo ConstructCopyStatement(CopyStmt *copyStatement, int64 shardId);
+static void SendCopyDataToPlacements(StringInfo dataBuffer,
+									 ShardConnections *shardConnections);
+static List * ConnectionList(HTAB *connectionHash);
+static void EndRemoteCopy(List *connectionList, bool stopOnFailure);
+static void ReportCopyError(PGconn *connection, PGresult *result);
+static void CopySendData(CopyOutState outputState, const void *databuf, int datasize);
+static void CopySendString(CopyOutState outputState, const char *str);
+static void CopySendChar(CopyOutState outputState, char c);
+static void CopySendInt32(CopyOutState outputState, int32 val);
+static void CopySendInt16(CopyOutState outputState, int16 val);
+static void CopyAttributeOutText(CopyOutState outputState, char *string);
+static inline void CopyFlushOutput(CopyOutState outputState, char *start, char *pointer);
+
+
+/*
+ * CitusCopyFrom implements the COPY table_name FROM ... for hash-partitioned
+ * and range-partitioned tables.
+ */
+void
+CitusCopyFrom(CopyStmt *copyStatement, char *completionTag)
+{
+	RangeVar *relation = copyStatement->relation;
+	Oid tableId = RangeVarGetRelid(relation, NoLock, false);
+	char *relationName = get_rel_name(tableId);
+	EState *executorState = NULL;
+	MemoryContext executorTupleContext = NULL;
+	ExprContext *executorExpressionContext = NULL;
+	List *shardIntervalList = NULL;
+	ListCell *shardIntervalCell = NULL;
+	char partitionMethod = '\0';
+	Var *partitionColumn = NULL;
+	HTAB *shardConnectionHash = NULL;
+	HASH_SEQ_STATUS status;
+	ShardConnections *shardConnections = NULL;
+	List *connectionList = NIL;
+	CopyState copyState = NULL;
+	TupleDesc tupleDescriptor = NULL;
+	uint32 columnCount = 0;
+	Datum *columnValues = NULL;
+	bool *columnNulls = NULL;
+	Relation rel = NULL;
+	ShardInterval **shardIntervalCache = NULL;
+	bool useBinarySearch = false;
+	TypeCacheEntry *typeEntry = NULL;
+	FmgrInfo *hashFunction = NULL;
+	FmgrInfo *compareFunction = NULL;
+	int shardCount = 0;
+	uint64 processedRowCount = 0;
+	ErrorContextCallback errorCallback;
+	CopyOutState copyOutState = NULL;
+	FmgrInfo *columnOutputFunctions = NULL;
```

Comment:
I agree, but the function cannot be easily split up (the while loop uses ~20 variables from the outer-context), and I don't think moving code into the PG_TRY that doesn't need to be there is a good idea. We could maybe get rid of some helper variables like columnCount and typeEntry by creating a few blocks or tiny functions, but I feel like that doesn't help readability much.